clicommand/util_keygen.go

```diff
@@ -0,0 +1,110 @@
+package clicommand
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/buildkite/agent/v3/internal/jwkutil"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/urfave/cli"
+	"golang.org/x/exp/slices"
+)
+
+type KeygenConfig struct {
+	Alg                  string `cli:"alg" validate:"required"`
+	KeyID                string `cli:"key-id" validate:"required"`
+	PrivateKeyFilename   string `cli:"private-key-filename" normalize:"filepath"`
+	PublicKeysetFilename string `cli:"public-keyset-filename" normalize:"filepath"`
+
+	LogLevel string `cli:"log-level"`
+	Debug    bool   `cli:"debug"`
+}
+
+var KeygenCommand = cli.Command{
+	Name:  "keygen",
+	Usage: "Generate a new key pair, used to sign and verify jobs",
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:   "alg",
+			EnvVar: "BUILDKITE_AGENT_KEYGEN_ALG",
+			Usage:  fmt.Sprintf("The JWS signing algorithm to use for the key pair. Valid algorithms are: %v", ValidSigningAlgorithms),
+		},
+		cli.StringFlag{
+			Name:   "key-id",
+			EnvVar: "BUILDKITE_AGENT_KEYGEN_PRIVATE_KEY_ID",
+			Usage:  "The ID to use for the private key",
+		},
+		cli.StringFlag{
+			Name:   "private-key-filename",
+			EnvVar: "BUILDKITE_AGENT_KEYGEN_PRIVATE_KEY_FILENAME",
+			Usage:  "The filename to write the private key to",
+		},
+		cli.StringFlag{
+			Name:   "public-keyset-filename",
+			EnvVar: "BUILDKITE_AGENT_KEYGEN_PUBLIC_KEYSET_FILENAME",
+			Usage:  "The filename to write the public keyset to",
+		},
+		DebugFlag,
+		LogLevelFlag,
+	},
+	Action: func(c *cli.Context) {
+		ctx := context.Background()
+		_, cfg, l, _, done := setupLoggerAndConfig[KeygenConfig](ctx, c)
```

Comment:
```suggestion
		ctx, cfg, l, _, done := setupLoggerAndConfig[KeygenConfig](context.Backgroud(), c)
```

We want people to use the context returned by the `setupLoggerAndConfig` method, as that has the experiments in it. When we check if an experiment is enabled, we return `false` if the context had no mention of experiments: https://github.com/buildkite/agent/blob/f69df3866b3b34078df053b29bcd83d20e203c7e/internal/experiments/experiments.go#L123-L127

This means we could mistakenly believe an experiment is disabled if we check the wrong context. To prevent such bugs, we should not expose the original context at all until it has been populated with experiments.

Reply:
This is just a particular example. In general, I think users will want to use the returned context in most situations.